File: Webserv/www/cgi-bin/cgi_upload.py
#!/usr/bin/env python3
import cgi
import os

UPLOAD_DIR = os.path.join(os.path.dirname(__file__), '..', 'upload')

print("Content-Type: text/html\r\n")

form = cgi.FieldStorage()
if "file" not in form:
    print("<html><body><h1>No file uploaded</h1></body></html>")
    exit()

fileitem = form["file"]
if not fileitem.filename:
    print("<html><body><h1>No file selected</h1></body></html>")
    exit()

filename = os.path.basename(fileitem.filename)
target_path = os.path.join(UPLOAD_DIR, filename)

try:
    with open(target_path, 'wb') as f:
        f.write(fileitem.file.read())
    print(f"<html><body><h1>File '{filename}' uploaded successfully!</h1></body></html>")
except Exception as e:
    print(f"<html><body><h1>Failed to upload file: {e}</h1></body></html>")

File: Webserv/www/cgi-bin/cgi_upload.php
<?php
#!/usr/bin/php
header("Content-Type: text/html; charset=utf-8");

// Check if a file was uploaded
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_FILES['file'])) {
    $uploadDir = dirname(__DIR__) . '/upload/';
    if (!is_dir($uploadDir)) {
        mkdir($uploadDir, 0777, true);
    }
    $filename = basename($_FILES['file']['name']);
    $targetPath = $uploadDir . $filename;

    if (move_uploaded_file($_FILES['file']['tmp_name'], $targetPath)) {
        echo "<html><body>";
        echo "<h2>Upload successful!</h2>";
        echo "<p>File saved as: " . htmlspecialchars($filename) . "</p>";
        echo "</body></html>";
    } else {
        echo "<html><body><h2>Upload failed!</h2></body></html>";
    }
} else {
    // Show upload form
    echo '<html><body>
        <form method="POST" enctype="multipart/form-data">
            <input type="file" name="file" />
            <input type="submit" value="Upload" />
        </form>
    </body></html>';
}
?>